src/utils/caccel.c

```diff
@@ -0,0 +1,310 @@
+	if (read != sizeof(in_header)) {
+		fprintf(stderr, "Unable to read header\n");
+		goto end_header_read;
+	}
+
+	if (parse_wave_header(&in_header, &info.channels, &info.in_rate,
+			      (unsigned int *)&info.in_format) == -1) {
+		fprintf(stderr, "Unable to parse header\n");
+		goto end_header_read;
+	}
+
+	info.in_blockalign = info.channels * in_header.fmt.samplebits / 8;
+	info.in_dmabuf_size = (DMA_BUF_SIZE / info.in_blockalign) * info.in_blockalign;
+
+	if (info.out_format == 0)
+		info.out_format = SNDRV_PCM_FORMAT_S16_LE;
+
+	info.out_samplebits = alsa_snd_pcm_format_width(info.out_format);
+	init_wave_header(&out_header, info.channels, info.out_rate, info.out_samplebits);
+
+	written = fwrite(&out_header, 1, sizeof(out_header), fd_dst);
+	if (written != sizeof(out_header)) {
+		fprintf(stderr, "Error writing output file header: %s\n",
+			strerror(errno));
+		goto end_header_read;
+	}
+
+	err = ioctl(fd_accel, SNDRV_COMPRESS_GET_CAPS, &caps);
+	if (err < 0) {
+		fprintf(stderr, "Error getting caps %d\n", err);
+		goto end_header_read;
+	}
+
+	codec_caps.codec = SND_AUDIOCODEC_PCM;
+	err = ioctl(fd_accel, SNDRV_COMPRESS_GET_CODEC_CAPS, &codec_caps);
+	if (err < 0) {
+		fprintf(stderr, "Error getting codec caps %d\n", err);
+		goto end_header_read;
+	}
+
+	params.buffer.fragment_size = DMA_BUF_SIZE;
+	params.buffer.fragments = 1;
+	params.codec.id = SND_AUDIOCODEC_PCM;
+	params.codec.ch_in  = info.channels;
+	params.codec.ch_out = info.channels;
+	params.codec.format = info.in_format;
+	params.codec.sample_rate = info.in_rate;
+	params.codec.pcm_format = info.out_format;
+	params.codec.options.src_d.out_sample_rate = info.out_rate;
+
+	err = ioctl(fd_accel, SNDRV_COMPRESS_SET_PARAMS, &params);
+	if (err < 0) {
+		fprintf(stderr, "Error setting params %d\n", err);
+		goto end_header_read;
+	}
+
+	err = ioctl(fd_accel, SNDRV_COMPRESS_TASK_CREATE, &task);
+	if (err < 0) {
+		fprintf(stderr, "Error creating task %d\n", err);
+		goto end_header_read;
+	}
+
+	bufin_start = mmap(NULL, MAP_BUF_SIZE, PROT_READ | PROT_WRITE,
+			   MAP_SHARED, task.input_fd, 0);
+	if (bufin_start == MAP_FAILED) {
+		fprintf(stderr, "Error mapping input buffer\n");
+		goto end_mmap_in;
+	}
+	memset(bufin_start, 0, MAP_BUF_SIZE);
+
+	bufout_start = mmap(NULL, MAP_BUF_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED,
+			    task.output_fd, 0);
+	if (bufout_start == MAP_FAILED) {
+		fprintf(stderr, "Error mapping output buffer\n");
+		goto end_mmap_out;
+	}
+	memset(bufout_start, 0, MAP_BUF_SIZE);
+
+	if (verbose)
+		printf("conversion is started\n");
+
+	status.seqno = task.seqno;
+
+	do {
+
+		read = fread(bufin_start, 1, info.in_dmabuf_size, fd_src);
+		if (read <= 0)
+			break;
+
+		task.input_size = read;
+
+		if (ioctl(fd_accel, SNDRV_COMPRESS_TASK_START, &task) < 0) {
+			fprintf(stderr, "Error starting task\n");
+			goto end_process_err;
+		}
+
+		if (ioctl(fd_accel, SNDRV_COMPRESS_TASK_STOP, &task.seqno) < 0) {
```

Comment:
STOP ioctl should not wait for result. STATUS or `poll()` should be used to wait for the task operation.

Reply:
yep